### PR TITLE
UNOMI-419 - on segment creation, validate segment condition and logging errors

### PR DESCRIPTION
**Pull request by giladw**

### CODE REVIEW

**Comment by giladw** [comment]
https://issues.apache.org/jira/browse/UNOMI-419

**Review by dgriffon — COMMENTED** [review]
Hello, thank you very much for your contribution to the project, this feature will really help.
I provided some suggestions, as suggestions, they are open to discussion. 

Thank you once again!

File: services/src/main/java/org/apache/unomi/services/impl/segments/SegmentServiceImpl.java [hunk 1]
```diff
@@ -233,14 +233,19 @@ public Segment getSegmentDefinition(String segmentId) {
...
+            persistenceService.validateCondition(segment.getCondition(), new Profile());
+            if (segment.getMetadata().isEnabled() && !segment.getMetadata().isMissingPlugins()) {
+                updateAutoGeneratedRules(segment.getMetadata(), segment.getCondition());
+            }
+            // make sure we update the name and description metadata that might not match, so first we remove the entry from the map
+            persistenceService.save(segment, null, true);
+            updateExistingProfilesForSegment(segment);
+        } catch (Exception e) {
+            logger.error("failed at setSegmentDefinition, condition={}", segment.getCondition().toString(), e);
+            throw e;
```
[the remark this answers is not in the document]
Comment (giladw) [reply services/src/main/java/org/apache/unomi/services/impl/segments/SegmentServiceImpl.java:247]: I agree, im removing the try-catch, and throwing if validation fails

File: services/src/main/java/org/apache/unomi/services/impl/segments/SegmentServiceImpl.java [hunk 2]
```diff
@@ -233,14 +233,19 @@ public Segment getSegmentDefinition(String segmentId) {
...
+        try {
+            ParserHelper.resolveConditionType(definitionsService, segment.getCondition());
+            persistenceService.validateCondition(segment.getCondition(), new Profile());
+            if (segment.getMetadata().isEnabled() && !segment.getMetadata().isMissingPlugins()) {
+                updateAutoGeneratedRules(segment.getMetadata(), segment.getCondition());
+            }
+            // make sure we update the name and description metadata that might not match, so first we remove the entry from the map
+            persistenceService.save(segment, null, true);
+            updateExistingProfilesForSegment(segment);
+        } catch (Exception e) {
```
[the remark this answers is not in the document]
Comment (giladw) [reply services/src/main/java/org/apache/unomi/services/impl/segments/SegmentServiceImpl.java:245]: ok, moving the log to the validation.

File: services/src/main/java/org/apache/unomi/services/impl/segments/SegmentServiceImpl.java [hunk 3]
```diff
@@ -233,14 +233,19 @@ public Segment getSegmentDefinition(String segmentId) {
     }
 
     public void setSegmentDefinition(Segment segment) {
-        ParserHelper.resolveConditionType(definitionsService, segment.getCondition());
-        if (segment.getMetadata().isEnabled() && !segment.getMetadata().isMissingPlugins()) {
-            updateAutoGeneratedRules(segment.getMetadata(), segment.getCondition());
+        try {
+            ParserHelper.resolveConditionType(definitionsService, segment.getCondition());
+            persistenceService.validateCondition(segment.getCondition(), new Profile());
```
[the remark this answers is not in the document]
Comment (giladw) [reply services/src/main/java/org/apache/unomi/services/impl/segments/SegmentServiceImpl.java:238]: yes, i will go with the boolean option

File: persistence-spi/src/main/java/org/apache/unomi/persistence/spi/PersistenceService.java [hunk 4]
```diff
@@ -305,6 +305,14 @@
...
     boolean testMatch(Condition query, Item item);
 
+    /**
+     * validates if a condition throws exception at query build.
+     *
+     * @param query the condition we're testing the specified item against
+     * @param item  the item we're checking against the specified condition
+     * @return {@code true} if the item satisfies the condition, {@code false} otherwise
+     */
+    void validateCondition(Condition query, Item item);
```
[the remark this answers is not in the document]
Comment (giladw) [reply persistence-spi/src/main/java/org/apache/unomi/persistence/spi/PersistenceService.java:315]: thanks, changing to boolean

**Review by dgriffon — APPROVED** [review]

**Pull request merged**